scrapy_poet/registry.py

```diff
@@ -0,0 +1,149 @@
+import logging
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from typing import Any, Callable, Dict, Iterable, Mapping, Optional, Type
+from warnings import warn
+
+from scrapy import Request
+from scrapy.crawler import Crawler
+from url_matcher import URLMatcher
+from web_poet import ItemPage, RulesRegistry
+from web_poet.rules import ApplyRule
+
+from scrapy_poet.utils._deprecation import _create_deprecated_class
+
+logger = logging.getLogger(__name__)
+
+
+class OverridesRegistryBase(ABC):
+    @abstractmethod
+    def overrides_for(self, request: Request) -> Mapping[Callable, Callable]:
+        """
+        Return a ``Mapping`` (e.g. a ``dict``) with type translation rules.
+        The key is the source type that is wanted to be replaced by
+        value, which is also a type.
+        """
+        pass
+
+
+class OverridesAndItemRegistry(OverridesRegistryBase, RulesRegistry):
```

Comment:
I'm thinking if there's a better name for this, such that it's general enough that we won't deprecate it again when a new feature comes out again, as well as it's distinct enough against `web_poet.RulesRegistry`.

Reply:
Given the import path gives additional context, I am fine with: `Registry`, `RuleRegistry`, `RulesRegistry` (a “registry of rules” is a “rule registry”, but for consistency with upstream this name may be best). Prepending `Scrapy` to any of those is also fine by me.

Reply:
Plot twist: scrapy-poet's registry has been removed in lieu of https://github.com/scrapinghub/web-poet/pull/112.

Ref: https://github.com/scrapinghub/scrapy-poet/pull/103/commits/4c1fa0283eba727c5dcb7c83a61969218ac97c07